src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs

```diff
@@ -72,19 +71,22 @@ public RazorPage()
         public DiagnosticSource DiagnosticSource { get; set; }
 
         /// <summary>
-        /// Gets the <see cref="TextWriter"/> that the page is writing output to.
+        /// Gets the <see cref="IHtmlContentBuilder"/> that the page is writing output to.
         /// </summary>
-        public virtual TextWriter Output
+        public virtual IHtmlContentBuilder Output
         {
             get
             {
-                if (ViewContext == null)
+                if (_contentBuilderScopes.Count > 0)
                 {
-                    var message = Resources.FormatViewContextMustBeSet("ViewContext", "Output");
-                    throw new InvalidOperationException(message);
+                    return _contentBuilderScopes.Peek();
                 }
 
-                return ViewContext.Writer;
+                return _originalWriter;
```

Comment:
We could. There are pieces of code that look at the stack to determine if they're inside a tag helper. Just makes them slightly obtuse now (`if (_contentBuilderScopes.Count > 1) // 1 is fine since it's the default writer`)